urlcaef331f4e_dreamable_cra_tgz-pJ8-coho_lp_project_LPProjectJ8/src/coho/common/number/ScaleType.java

```diff
@@ -0,0 +1,309 @@
+package coho.common.number;
+
+
+public class ScaleType extends BasicType{
+	public ScaleType(Class classType, ScaleNumber zero, ScaleNumber one){
+		super(classType,zero,one);
+	}
+	public ScaleNumber zero(){
+		return (ScaleNumber)zero;
+	}
+	public ScaleNumber one(){
+		return (ScaleNumber)one;
+	}
+	
+	/*
+	 * Promotion data
+	 * Promotion rule: 	1)boolean->int->double->apr.
+	 * 					2)basic->interval 
+	 */
+	private static ScaleType[] types = null;
+	private static int typeKey(ScaleType x){
+		if(types==null){
+			types = new ScaleType[] {CohoBoolean.type, 
+					CohoInteger.type, CohoLong.type, CohoDouble.type, CohoAPR.type};
+		}
+		for(int i=0; i<types.length; i++){
+			if(x==types[i])
+				return i;
+		}
+		throw new IllegalArgumentException("unknown CohoType:  " + x.name());
+	}
+
+	/**
+	 * This function depends on the order of types to work correctly;
+	 */
+	public ScaleType promote(ScaleType that) {
+		if(this==that){
+			return this;
+		}
+		int tk1 = typeKey(this), tk2 = typeKey(that);
+		int tk = Math.max(tk1,tk2);
+		return types[tk];
+	}
+
+	public static ScaleNumber[] promote(ScaleNumber a, ScaleNumber b){
+		ScaleType t = (a.type()==b.type())?a.type():a.type().promote(b.type());
+		return new ScaleNumber[]{promote(a,t),promote(b,t)};
+	}
+	public static ScaleNumber[] promote(ScaleNumber a, Number b){
+		return promote(a,promote(b));
+	}
+
+
+	private static ScaleNumber promote(ScaleNumber b, ScaleType t){
+		if(t == b.type()){
+			return b;
+		}
+		if(t == CohoBoolean.type)
+			return CohoBoolean.create(b);
+		else if(t == CohoInteger.type)
+			return CohoInteger.create(b);
+		else if(t == CohoLong.type)
+			return CohoLong.create(b);
+		else if(t == CohoDouble.type)
+			return CohoDouble.create(b);
+		else if(t == CohoAPR.type)
+			return CohoAPR.create(b);
+		else
+			throw new IllegalArgumentException("Unknown subclass of CohoNumber: " + t);
+	}
+	public static ScaleNumber promote(Number x){
+		if( x instanceof Integer || x instanceof Short || x instanceof Byte){
+			return CohoInteger.create(x);
+		}else if(x instanceof Long){
+			return CohoLong.create(x);
+		}else if(x instanceof Float || x instanceof Double){
+			return CohoDouble.create(x);
+		}else{
+			return CohoDouble.create(x.doubleValue());
+		}
+	}
+	
+	
+	public static ScaleNumber promoteOp(ScaleNumber n1, ScaleNumber n2, CohoNumber.ArithOp op){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 104 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.